### PR TITLE
Refactor settings menu in swiftui only UI

### PR DESCRIPTION
**Pull request by liyafly**
#286  初步完成UI，功能是否需要等使用swift重构后添加还是使用桥接Objc完成呢？

### CODE REVIEW

**Comment by tisfeng** [comment]
一些简单的功能，可以在重构 UI 时顺便也重写了。一些比较复杂的，像输入翻译这几个，可以先桥接使用 objc，后面专门弄个 PR 来重写。

**Comment by tisfeng** [comment]
目前这个【检查更新】功能好像不对，麻烦你也看一下。

File: Easydict/NewApp/View/MenuItemView.swift [hunk 1]
```diff
@@ -8,16 +8,27 @@
 
 import Sparkle
 import SwiftUI
+import ZipArchive
 
 @available(macOS 13, *)
 struct MenuItemView: View {
     var body: some View {
+        // ️.menuBarExtraStyle为 .menu 时某些控件可能会失效 ，只能显示内容（按照菜单项高度、图像以 template 方式渲染）无法交互 ，比如 Stepper、Slider 等，像基本的 Button、Text、Divider、Image 等还是能正常显示的。
+        // Button 和Label的systemImage是不会渲染的
```
Comment (tisfeng) [inline Easydict/NewApp/View/MenuItemView.swift:17]: 这个是什么意思，没办法修改默认的菜单项高度吗？我希望增加一点高度，和之前保持一致。

image

image

**Review by tisfeng — CHANGES_REQUESTED** [review]

File: Easydict/NewApp/View/MenuItemView.swift (hunk 1 above)
Comment (liyafly) [reply Easydict/NewApp/View/MenuItemView.swift:17]: 高度，还不确定，只是对于   Button("Sign In", systemImage: "arrow.up", action: signIn) 这种的，不回显示systemImage 图片。高度的话我一会尝试一下。

Comment (liyafly) [reply Easydict/NewApp/View/MenuItemView.swift:17]: > 这个是什么意思，没办法修改默认的菜单项高度吗？我希望增加一点高度，和之前保持一致。
> 
> 
> 
> image
> 
> 
> 
> image
> 
> 

我尝试了 只能使用.window才能自定义高度， 这个高度想要多高呢？ 我抽时间在对UI进行一些修改 同时之前的settinglink 会有一个默认的图标在macOS14上，这个默认的图标需要吗？

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:17]: 和之前的效果差不多就行，设置的图标不需要。

Comment (liyafly) [reply Easydict/NewApp/View/MenuItemView.swift:17]: > 和之前的效果差不多就行，设置的图标不需要。

可以的，但这个PR 的改动可能晚几天完成，需要关了之后我在重新提吗？还是先放着？

Comment (phlpsong) [reply Easydict/NewApp/View/MenuItemView.swift:17]: Yes, just noticed that 😅

Sure, will take a look later.

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:17]: 高度可以放后面再优化，你可以先完成这个 PR 的内容，我们先合并，争取早点用 SwiftUI 替换设置页。

Comment (phlpsong) [reply Easydict/NewApp/View/MenuItemView.swift:17]: After some research, it looks like a macOS bug([Sonoma Issue on NSMenuItem keyEquivalent value](https://stackoverflow.com/questions/77210652/sonoma-issue-on-nsmenuitem-keyequivalent-value)), currently I could not find a perfect solution for this. 

I tried to set the image size programmatically but also has side effects based on my test result. 

My suggestion is not to increase the menu item height, and to use the system's default value. 
I checked other apps like `Raycast` & `EuDic` & `ClashX` on my Mac device, and all of these use default height.

Screenshot 2024-01-07 at 21 47 36

Screenshot 2024-01-07 at 21 55 26



@tisfeng  What do you think? Please let me know your thoughts.

Comment (liyafly) [reply Easydict/NewApp/View/MenuItemView.swift:23]: 是的，如果使用 window 倒是可以自定义高度，缺点是样式需要完全自定义。目前menu的高度确实不可更改。

**Comment by liyafly** [comment]
我已经添加objc桥接，但debug下部分功能系统权限授权有问题，无法做到全面测试。请review

File: Easydict/NewApp/View/MenuItemView.swift (hunk 1 above)
Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:23]: 行，那先就使用默认高度。

Comment (phlpsong) [reply Easydict/NewApp/View/MenuItemView.swift:23]: @tisfeng  If you are fine with that, I can submit a PR to remove the `increaseMenuItemHeight` functions.

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:23]: 行，你另外提个 PR 删了吧。

**Comment by tisfeng** [comment]
有两个小问题：

- @property window 出现警告，可能是桥接文件添加 #import "EZWindowManager.h" 导致。
-  Sparkle 有 gentle reminders 警告，我之前是通过实现 supportsGentleScheduledUpdateReminders 消除了这个警告。

image

**Comment by tisfeng** [comment]
> 我已经添加objc桥接，但debug下部分功能系统权限授权有问题，无法做到全面测试。请review

参考下面，你可以在 Easydict-debug.xcconfig 配置文件添加你自己的开发者账号，方便后续调试。


```
DEVELOPMENT_TEAM = 79NQA2XYHM
CODE_SIGN_IDENTITY = Apple Development
CODE_SIGN_STYLE = Manual
```

由于这个文件不应该提交，可以使用 git 忽略这个文件变动。（后面假如切换分支遇到问题，可以再改回来）

```
git update-index --skip-worktree Easydict-debug.xcconfig
```

File: Easydict/NewApp/View/MenuItemView.swift (hunk 1 above)
Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:23]: @phlpsong 这个链接打不开，是图片吗还是什么？

> I tried to set the image size programmatically but also has side effects based on my test result.

我没有试过，请问这里修改图片大小，会导致什么问题？

image

**Comment by liyafly** [comment]
> 有两个小问题：
> 
> * @Property window 出现警告，可能是桥接文件添加 #import "EZWindowManager.h" 导致。
> * Sparkle 有 gentle reminders 警告，我之前是通过实现 supportsGentleScheduledUpdateReminders 消除了这个警告。
> 
> image

好的，我下周进行修改。thanks

File: Easydict/NewApp/View/MenuItemView.swift (hunk 1 above)
Comment (phlpsong) [reply Easydict/NewApp/View/MenuItemView.swift:23]: If only apply workaround to no image item

```
if (!item.image) {
        NSLog(@"%@", item.title);
        NSImage *image = [[NSImage alloc] initWithSize:NSMakeSize(1, lineHeight)];
        [item setImage:image];
 }
```
result: 
Screenshot 2024-01-07 at 23 16 37

I also tried to customize/resize the image size but not work either.

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:23]: 我尝试设置设置图片，修改图片大小，好像看起来没什么问题？

```objc
/// Increase menu item height. Ref: https://stackoverflow.com/questions/18031666/nsmenuitem-height
- (void)increaseMenuItemHeight:(NSMenuItem *)item lineHeightRatio:(CGFloat)lineHeightRatio {
    NSFont *font = [NSFont systemFontOfSize:[NSFont systemFontSize]];
    CGFloat fontLineHeight = (font.ascender + fabs(font.descender));
    CGFloat lineHeight = fontLineHeight * lineHeightRatio;
    // Ref stackoverflow: https://stackoverflow.com/a/18034142/8378840 
//    NSImage *image = [[NSImage alloc] initWithSize:NSMakeSize(1, lineHeight)];
    
    NSImage *image = [NSImage imageNamed:@"error"];
    image.size = CGSizeMake(30, 30);
    [item setImage:image];
}
```
image

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:23]: 试了一下，发现是图片类型问题，普通的图片没问题，系统的 SF 图片这里修改 image.size 无效 😓

```
image = [NSImage imageWithSystemSymbolName:@"keyboard" accessibilityDescription:nil];
image.size = CGSizeMake(30, 30);
```

Comment (phlpsong) [reply Easydict/NewApp/View/MenuItemView.swift:23]: Yes, no effect on system image. I also tried to resize the image, but it looks wired and could not remove the leading space of other menu.

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:23]: 已合并，去掉了高度改动 https://github.com/tisfeng/Easydict/pull/318

File: Easydict/NewApp/View/MenuItemView.swift [hunk 2]
```diff
@@ -8,16 +8,33 @@
 
 import Sparkle
 import SwiftUI
+import ZipArchive
 
 @available(macOS 13, *)
 struct MenuItemView: View {
+    private let updater: SPUUpdater
```
Comment (Kyle-Ye) [inline Easydict/NewApp/View/MenuItemView.swift:15]: Do we need to add `ObservableObject` here?

```swift
final class MenuItemStore: ObservableObject {
	let update: SPUUpdater
	init(updater: SPUUpdater) {
	}
}
```

```
struct MenuItemView: View {
	@StateObject var store = MenuItemStore()
}
```

Comment (liyafly) [reply Easydict/NewApp/View/MenuItemView.swift:15]: This is a very good method, and the official code examples are similar to this. However, this only checks for updates and does not involve listening to more status changes and UI changes. Of course, we need to listen to @tisfeng 
 suggestions on this. Personally, I think we can maintain the status.

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:15]: I'm not very familiar with Swift/SwiftUI, and that part of my knowledge is still a long way off. However, it does look like there is no need to listen for state changes here.

Comment (tisfeng) [reply Easydict/NewApp/View/MenuItemView.swift:15]: I'd like to rewrite the code in such a way that the functionality should just be as consistent as possible with what it was before.

Also, I made some changes to [Check for Updates] before, and the `feedURLString` in the debug environment is a convenient URL for local debugging, which I'd like to add as well. @liyafly 

https://github.com/tisfeng/Easydict/blob/261b625f3d770365ff1fa0ddabe6a030b0592d08/Easydict/App/AppDelegate.m#L104

Comment (liyafly) [reply Easydict/NewApp/View/MenuItemView.swift:15]: Understood, I’ve grasped your requirements. I also aim to refactor the code into a more scalable architecture, while carefully comparing it with the previous functionalities to ensure a proper restructuring. I will complete the initial development within this week.

**Comment by liyafly** [comment]
> 有两个小问题：
> 
> 
> 
> - @property window 出现警告，可能是桥接文件添加 #import "EZWindowManager.h" 导致。
> 
> -  Sparkle 有 gentle reminders 警告，我之前是通过实现 supportsGentleScheduledUpdateReminders 消除了这个警告。
> 
> 
> 
> image
> 
> 

I added the nullable modifier to `@property (nullable, nonatomic, weak) EZBaseQueryWindow *window`;, but Xcode is still throwing a warning. I’m not quite sure why this is happening. Do you have any insights on this issue?

**Comment by tisfeng** [comment]
Wait for me to take a look.

**Comment by tisfeng** [comment]
It's strange, I've tried a number of ways and none of them work. This is caused by being referenced by a Swift bridge file, and I'm not familiar with Swift.

If there's really no way to fix it, we can just ignore the warning:

```objc
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wnullability"
@property (nonatomic, weak) EZBaseQueryWindow *window;
#pragma clang diagnostic pop
```

**Comment by Kyle-Ye** [comment]
> It's strange, I've tried a number of ways and none of them work. This is caused by being referenced by a Swift bridge file, and I'm not familiar with Swift.
> 
> If there's really no way to fix it, we can just ignore the warning:
> 
> ```objc
> #pragma clang diagnostic push
> #pragma clang diagnostic ignored "-Wnullability"
> @property (nonatomic, weak) EZBaseQueryWindow *window;
> #pragma clang diagnostic pop
> ```
> 
> Ref: [stackoverflow.com/a/39918786/8378840](https://stackoverflow.com/a/39918786/8378840)

Nope. This is just a small issue and we should fix it easily.

The log message already tells us how to fix it.
```shell
/Users/kyle/Github/Easydict/Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.h:21:48: warning: conflicting nullability specifier on return types, '_Nullable' conflicts with existing specifier 'nonnull' [-Wnullability]
@property (nonatomic, weak) EZBaseQueryWindow *window;
                                               ^
In file included from /Users/kyle/Github/Easydict/Easydict/Feature/PerferenceWindow/EZSettingViewController.m:12:
/Users/kyle/Github/Easydict/Easydict/Feature/Utility/EZCategory/NSViewController+EZWindow.h:15:1: note: previous declaration is here
- (NSWindow *)window;
^
1 warning generated.
```

`weak` just infer `nullable` implicitly. I believe the right fix is to remove `NS_ASSUME_NONNULL_BEGIN` and `NS_ASSUME_NONNULL_END` in `NSViewController+EZWindow.h` or add a nullable mark here(Suggested).

```ObjectiveC
NS_ASSUME_NONNULL_BEGIN

@interface NSViewController (EZWindow)

- (NSWindow *)window;

@end

NS_ASSUME_NONNULL_END
```

```diff
NS_ASSUME_NONNULL_BEGIN

@interface NSViewController (EZWindow)

-- (NSWindow *)window;
+- (nullable NSWindow *)window;
@end

NS_ASSUME_NONNULL_END
```

**Comment by tisfeng** [comment]
Awesome, I thought it was warning me about `@property (nonatomic, weak) EZBaseQueryWindow *window;`, didn't look at the log details carefully 😥

image

image

**Comment by liyafly** [comment]
> > It's strange, I've tried a number of ways and none of them work. This is caused by being referenced by a Swift bridge file, and I'm not familiar with Swift.
> > If there's really no way to fix it, we can just ignore the warning:
> > ```objc
> > #pragma clang diagnostic push
> > #pragma clang diagnostic ignored "-Wnullability"
> > @property (nonatomic, weak) EZBaseQueryWindow *window;
> > #pragma clang diagnostic pop
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Ref: [stackoverflow.com/a/39918786/8378840](https://stackoverflow.com/a/39918786/8378840)
> 
> Nope. This is just a small issue and we should fix it easily.
> 
> The log message already tells us how to fix it.
> 
> ```shell
> /Users/kyle/Github/Easydict/Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.h:21:48: warning: conflicting nullability specifier on return types, '_Nullable' conflicts with existing specifier 'nonnull' [-Wnullability]
> @property (nonatomic, weak) EZBaseQueryWindow *window;
>                                                ^
> In file included from /Users/kyle/Github/Easydict/Easydict/Feature/PerferenceWindow/EZSettingViewController.m:12:
> /Users/kyle/Github/Easydict/Easydict/Feature/Utility/EZCategory/NSViewController+EZWindow.h:15:1: note: previous declaration is here
> - (NSWindow *)window;
> ^
> 1 warning generated.
> ```
> 
> `weak` just infer `nullable` implicitly. I believe the right fix is to remove `NS_ASSUME_NONNULL_BEGIN` and `NS_ASSUME_NONNULL_END` in `NSViewController+EZWindow.h` or add a nullable mark here(Suggested).
> 
> ```objc
> NS_ASSUME_NONNULL_BEGIN
> 
> @interface NSViewController (EZWindow)
> 
> - (NSWindow *)window;
> 
> @end
> 
> NS_ASSUME_NONNULL_END
> ```
> 
> ```diff
> NS_ASSUME_NONNULL_BEGIN
> 
> @interface NSViewController (EZWindow)
> 
> -- (NSWindow *)window;
> +- (nullable NSWindow *)window;
> @end
> 
> NS_ASSUME_NONNULL_END
> ```
> 
> You can just merge the patch here [liyafly#1](https://github.com/liyafly/Easydict/pull/1).

Okay, I didn't see the whole log either, but according to you, does this duplicate the attribute definition in the category @tisfeng

**Comment by tisfeng** [comment]
No, they are not the same. I didn't notice that they had the same name before. Maybe we can rename the name of `window` to `baseQueryWindow`, like this

```objc
@property (nonatomic, weak) EZBaseQueryWindow *baseQueryWindow;
```

**Comment by liyafly** [comment]
> No, they are not the same. I didn't notice that they had the same name before. Maybe we can rename the name of `window` to `baseQueryWindow`, like this
> 
> ```objc
> @property (nonatomic, weak) EZBaseQueryWindow *baseQueryWindow;
> ```

“Certainly, after making such changes, it’s still necessary to add nullable. I’ll go through everything to ensure that all the renaming is correct
```objc
@property (nullable, nonatomic, weak) EZBaseQueryWindow *baseQueryWindow;
 ```

**Comment by tisfeng** [comment]
I tried renaming the `window` to `baseQueryWindow` and it looks like the warning is resolved?

**Comment by liyafly** [comment]
Yes, It would be nice to rename it, because weak just infer nullable, and it would be clearer to add nullable, don't you think? @tisfeng

**Comment by tisfeng** [comment]
> Yes, It would be nice to rename it, because weak just infer nullable, and it would be clearer to add nullable, don't you think? @tisfeng

Yes, I just want to confirm that renaming the property can fix this warning, and also explicitly adding nullable is ok.

**Comment by tisfeng** [comment]
I ran the code, and there was no problem. 

If there are no other questions, we can continue to review.

**Comment by liyafly** [comment]
> I ran the code, and there was no problem. 
> 
> 
> 
> If there are no other questions, we can continue to review.

I'm not too familiar with SwiftUI, so feel free to point out any code issues — I'm here to learn. I've tested the features extensively and they seem fine for now. Welcome to review and give feedback! 

My English isn't very good, please bear with me.

**Review by tisfeng — APPROVED** [review]

**Review by CanglongCl — APPROVED** [review]

**Pull request merged**